scipy/stats/_rcont/_rcont.c

```diff
@@ -0,0 +1,242 @@
+
+  // shuffle work
+  for (int i = n - 1; i > 0; --i)
+  {
+    int j = random_interval(rstate, i);
+    int tmp = work[j];
+    work[j] = work[i];
+    work[i] = tmp;
+  }
+
+  // clear table
+  for (int i = 0, nrc = (nr * nc); i < nrc; ++i)
+    matrix[i] = 0;
+
+  // fill table
+  for (int ir = 0; ir < nr; ++ir)
+  {
+    int ri = (int)r[ir];
+    while (ri--)
+      *ptr(matrix, nr, nc, ir, *work++) += 1;
+  }
+}
+
+/*
+  Generate random two-way table with given marginal totals.
+
+  Patefield's algorithm adapted from AS 159 Appl. Statist. (1981) 91-97. This
+  algorithm has O(K log(N)) complexity in time for a table with K cells and N
+  entries in total. It requires only a small constant stack space.
+
+  The original FORTRAN code was hand-translated to C. Changes to the original:
+
+  - The computation of a look-up table of log-factorials was replaced with
+    calls to lgamma, which lifts the limitation that the code only works for
+    tables with less then 5000 entries.
+  - The data type of input and output arrays was changed to double to minimize
+    type conversions. Users are responsible for passing only integral numbers.
+  - The original implementation allocated a column vector JWORK, but this is
+    not necessary. The vector can be folded into the last column of the output
+    matrix.
+  - The function uses Numpy's random number generator and distribution library.
+  - The algorithm now handles zero entries in row or column vector. When a
+    zero is encountered, the output matrix is filled with zeros along that row
+    or column and the algorithm proceeds to the next entry.
+
+  The argument ntot is used to detect whether the function has been run before
+  and has to be zero initialised.
+*/
+void rcont2(double *matrix, int nr, const double *r, int nc, const double *c,
+            double ntot, bitgen_t *rstate)
+{
+  // jwork is folded into matrix using last row
+  double *jwork = ptr(matrix, nr, nc, nr - 1, 0);
+  for (int i = 0; i < nc; ++i)
+  {
+    jwork[i] = c[i];
+  }
+
+  double jc = ntot;
+  double ib = 0;
+  // last row is not random due to constraint
+  for (int l = 0; l < nr - 1; ++l)
+  {
+    double ia = r[l]; // first term
+    if (ia == 0)
+    {
+      for (int i = 0; i < nc; ++i)
+        *ptr(matrix, nr, nc, l, i) = 0;
+      continue;
+    }
+    double ic = jc; // second term
+    jc -= r[l];
+    // last column is not random due to constraint
+    for (int m = 0; m < nc - 1; ++m)
+    {
+      const double id = jwork[m]; // third term
+      const double ie = ic;       // eight term
+      ic -= id;
+      ib = ie - ia;
+      // must be after ib calculation, which is used at the end
+      if (c[m] == 0)
+      {
+        for (int i = 0; i < nr; ++i)
+          *ptr(matrix, nr, nc, i, m) = 0;
+        continue;
+      }
+      const double ii = ib - id; // forth term
+      if (ie == 0)
+      {
+        for (int j = m; j < nc - 1; ++j)
+          *ptr(matrix, nr, nc, l, j) = 0;
+        ia = 0;
```

Comment:
This cannot be removed. After line 177, we break out of the inner loop and arrive at line 204, where ia is assigned to the matrix.

Reply:
Oh, OK, I see what's happening. This code wasn't covered which is why I too was unsure whether it would work. It still shows that this part is uncovered. Could you come up with an input that covers this case and add it as a test?

Reply:
I am trying. It is not obvious.

Reply:
I tried, but I was not successful so far in constructing test cases that cover 100 % of the C code. We can still do this later.

Reply:
Thanks for trying! Yes, we can add more tests later. I don't think this is necessary to do right now.